In [79]:
import pandas as pd
pd.set_option('display.max_colwidth', 100)
import numpy as np
import gseapy as gp

# names = gp.get_library_name()
gobp = 'GO_Biological_Process_2021'
gomf = 'GO_Molecular_Function_2021'
gocc = 'GO_Cellular_Component_2021'

d = pd.read_csv(r'fdr5.csv', index_col=0)
g = d.loc[d['PPDE'] > 0.95].sort_values('PPDE', ascending=False).iloc[:1000].sort_index()

r = gp.enrichr(
    gene_list=g.index.to_list(), 
    gene_sets=[gocc, gomf, gobp],       
    cutoff=0.05,
    organism='Human',
    ).results

# I used the data from part 1 of homework 3 that uses only PBMC data. 

### 1. Top 3 GO Bio Processes

In [82]:
r.loc[r['Gene_set'] == 'GO_Biological_Process_2021'].iloc[0:3,[1, 4]]

Term  Adjusted P-value
933                        aerobic electron transport chain (GO:0019646)      1.884799e-16
934  mitochondrial ATP synthesis coupled electron transport (GO:0042775)      1.884799e-16
935    mitochondrial electron transport, NADH to ubiquinone (GO:0006120)      1.313649e-09

### 2. Top 3 GO Molecular Functions

In [89]:
r.loc[r['Gene_set'] == 'GO_Molecular_Function_2021'].iloc[0:3,[1, 4]]

Term  Adjusted P-value
301  oxidoreduction-driven active transmembrane transporter activity (GO:0015453)      5.046719e-15
302                            NADH dehydrogenase (quinone) activity (GO:0050136)      4.727279e-10
303                         NADH dehydrogenase (ubiquinone) activity (GO:0008137)      4.727279e-10

### 3. Top 3 GO Cellular Components

In [90]:
r.loc[r['Gene_set'] == 'GO_Cellular_Component_2021'].iloc[0:3,[1, 4]]

Term  Adjusted P-value
0        mitochondrial membrane (GO:0031966)      4.288697e-23
1  mitochondrial inner membrane (GO:0005743)      3.534770e-21
2      organelle inner membrane (GO:0019866)      2.588383e-19

## 4. Does the selected GO term make sense given the list of DE genes?



### Look at GO term aerobic electron transport chain (GO:0019646)

- This makes sense based on the name. Covid sufferers have trouble turning over oxygen, so it would make sense that RNA transcripts related to aerobic oxidation in the ETC will be DE. Likely these genes will be highly up-regulated because the cell is trying to boost these metabolic pathways to provide enough oxygen for the cells.
<br>  
- A google scholar search for `"covid-19"  aerobic "electron transport chain"` with year filter since 2021 yields 601 results which suggests that at least electron transport chain and covid-19 are strongly correlated.
<br>  

- This paper https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7934846/ states the following: 
>Our observations demonstrate that, within 7 days of infection, morbidity, wasting, and failure to thrive in these animals is associated with significant decrease in expression of genes related to key cellular metabolic processes regulating aerobic cellular respiration and energy production.
<br> 

However, their statement that expression of genes decreases actually disagrees with my hypothesis above. There is some effect from SARS-CoV-2 that causes the cells to express these important proteins at a much lower level.

In [65]:
cc = r.loc[r['Gene_set'] == 'GO_Cellular_Component_2021']
bp = r.loc[r['Gene_set'] == 'GO_Biological_Process_2021']
mf = r.loc[r['Gene_set'] == 'GO_Molecular_Function_2021']

['GO_Cellular_Component_2021' 'GO_Molecular_Function_2021'
 'GO_Biological_Process_2021']


In [66]:
cc

Gene_set  \
0    GO_Cellular_Component_2021   
1    GO_Cellular_Component_2021   
2    GO_Cellular_Component_2021   
3    GO_Cellular_Component_2021   
4    GO_Cellular_Component_2021   
..                          ...   
296  GO_Cellular_Component_2021   
297  GO_Cellular_Component_2021   
298  GO_Cellular_Component_2021   
299  GO_Cellular_Component_2021   
300  GO_Cellular_Component_2021   

                                                  Term  Overlap       P-value  \
0                  mitochondrial membrane (GO:0031966)   85/469  1.424816e-25   
1            mitochondrial inner membrane (GO:0005743)   67/328  2.348684e-23   
2                organelle inner membrane (GO:0019866)   66/346  2.579784e-21   
3                    mitochondrial matrix (GO:0005759)   57/348  1.905222e-15   
4    mitochondrial respiratory chain complex I (GO:...    17/42  5.200591e-12   
..                                                 ...      ...           ...   
296                platelet alpha granule (GO:0031091)     1/90  9.902133e-01   
297  integral component of plasma membrane (GO:0005...  55/1454  9.906198e-01   
298                    cell-cell junction (GO:0005911)    6/271  9.938035e-01   
299                           Golgi lumen (GO:0005796)    1/100  9.941539e-01   
300                     neuron projection (GO:0043005)   16/556  9.953697e-01   

     Adjusted P-value  Old P-value  Old Adjusted P-value  Odds Ratio  \
0        4.288697e-23            0                     0    4.503529   
1        3.534770e-21            0                     0    5.155836   
2        2.588383e-19            0                     0    4.724381   
3        1.433680e-13            0                     0    3.886161   
4        2.608963e-10            0                     0   13.126144   
..                ...          ...                   ...         ...   
296      9.953697e-01            0                     0    0.212696   
297      9.953697e-01            0                     0    0.732235   
298      9.953697e-01            0                     0    0.426749   
299      9.953697e-01            0                     0    0.191110   
300      9.953697e-01            0                     0    0.555857   

     Combined Score                                              Genes  
0        257.649535  COX7B;MRPS15;NDUFA13;MPV17L2;MRPL36;MRPL37;COX...  
1        268.647918  COX7B;MRPS15;NDUFA13;MPV17L2;MRPL36;MRPL37;COX...  
2        223.966731  COX7B;MRPS15;NDUFA13;MPV17L2;MRPL36;MRPL37;COX...  
3        131.718220  GFM1;ACAA2;MRPL37;TXN2;CA5B;ME2;IDH3A;BCKDHA;M...  
4        341.046752  NDUFA9;NDUFA13;NDUFA7;NDUFA6;NDUFB5;NDUFA4;NDU...  
..              ...                                                ...  
296        0.002092                                                HGF  
297        0.006901  CNTNAP1;GPR68;PTGDR2;KCNC4;PTPRO;ASGR1;MILR1;F...  
298        0.002653            ARVCF;CTNND1;NECTIN3;FGFRL1;AJAP1;ACTG1  
299        0.001121                                              DEFA1  
300        0.002580  MAGEE1;PTGDR2;KCNC4;PTPRO;ADAM22;SLC1A3;ANK3;C...  

[301 rows x 10 columns]

In [22]:
df = pd.read_csv('https://raw.githubusercontent.com/kuanrongchan/vaccine-studies/main/Ad5_seroneg.csv',index_col=0)
DEGs_up_1d = (df[(df['fc_1d'] > 1.5) & (df['qval_1d'] < 0.05)]).index.tolist()
DEGs_down_1d = (df[(df['fc_1d'] < -1.5) & (df['qval_1d'] < 0.05)]).index.tolist()

In [44]:
enr_GOBP_up = gp.enrichr(gene_list=DEGs_up_1d ,
 gene_sets=['GO_Biological_Process_2021'],
 organism='Human', 
 description='DEGs_up_1d',
 outdir='test/enr_DEGs_GOBP_up',
 cutoff=0.05 
 )

In [45]:
enr_GOBP_up.results

Gene_set  \
0     GO_Biological_Process_2021   
1     GO_Biological_Process_2021   
2     GO_Biological_Process_2021   
3     GO_Biological_Process_2021   
4     GO_Biological_Process_2021   
...                          ...   
3688  GO_Biological_Process_2021   
3689  GO_Biological_Process_2021   
3690  GO_Biological_Process_2021   
3691  GO_Biological_Process_2021   
3692  GO_Biological_Process_2021   

                                                   Term  Overlap  \
0      cytokine-mediated signaling pathway (GO:0019221)  102/621   
1             defense response to symbiont (GO:0140546)   45/124   
2                defense response to virus (GO:0051607)   46/133   
3     cellular response to interferon-gamma (GO:0071...   39/121   
4     cellular response to type I interferon (GO:007...    29/65   
...                                                 ...      ...   
3688                rRNA metabolic process (GO:0016072)    1/162   
3689                       gene expression (GO:0010467)    5/356   
3690        mRNA splicing, via spliceosome (GO:0000398)    3/274   
3691                   ribosome biogenesis (GO:0042254)    1/192   
3692                           translation (GO:0006412)    1/214   

           P-value  Adjusted P-value  Old P-value  Old Adjusted P-value  \
0     4.038333e-34      1.491357e-30            0                     0   
1     7.986735e-31      1.474751e-27            0                     0   
2     2.078736e-30      2.558924e-27            0                     0   
3     1.174254e-24      1.084130e-21            0                     0   
4     2.317999e-23      1.426729e-20            0                     0   
...            ...               ...          ...                   ...   
3688  9.989039e-01      9.997476e-01            0                     0   
3689  9.990978e-01      9.997476e-01            0                     0   
3690  9.992062e-01      9.997476e-01            0                     0   
3691  9.996896e-01      9.998758e-01            0                     0   
3692  9.998758e-01      9.998758e-01            0                     0   

      Odds Ratio  Combined Score  \
0       5.100533      392.190496   
1      14.020292      971.639286   
2      13.025317      890.225617   
3      11.614731      639.988034   
4      19.470644     1014.785675   
...          ...             ...   
3688    0.144054        0.000158   
3689    0.328683        0.000297   
3690    0.255885        0.000203   
3691    0.121236        0.000038   
3692    0.108588        0.000013   

                                                  Genes  
0     CD86;IL1RN;CSF1;CD80;IFIT5;IL1RAP;IFI35;CXCL13...  
1     RTP4;CD40;RNASEL;IFIT5;IFI6;IFIT1;IFI44L;IFIT3...  
2     RTP4;CD40;RNASEL;IFIT5;IFI6;IFIT1;IFI44L;IFIT3...  
3     CCL13;SP100;PTAFR;ICAM1;OASL;CCL8;CCL7;TRIM5;S...  
4     RNASEL;SP100;IFIT5;IFI6;ADAR;IFI35;IFIT1;IFIT3...  
...                                                 ...  
3688                                               XRN1  
3689                      RNASEL;NUP62;ADAR;NUP43;NUP58  
3690                                 SMU1;METTL14;CSTF3  
3691                                               GLUL  
3692                                             GTPBP1  

[3693 rows x 10 columns]